### PR TITLE
Hashmap repeated word

### PR DESCRIPTION
**Pull request by chrisyamas**

### CODE REVIEW

**Comment by chrisyamas** [comment]
- [x] Top-level README “Table of Contents” is updated
 - [x] README for this challenge is complete
       - [x] Summary, Description, Approach & Efficiency, Solution
       - [x] Picture of whiteboard
       - [x] Link to code
 - [x] Feature tasks for this challenge are completed
 - [x] Unit tests written and passing
       - [x] “Happy Path” - Expected outcome
       - [x] Expected failure
       - [x] Edge Case (if applicable/obvious)

File: python/code_challenges/hashtable_repeated_word.py [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
-    pass
+def first_repeated_word(string):
+    text_no_symbols = re.sub(r'[^\w]', ' ', string)
+    text_one_space = re.sub('\\s+', ' ', text_no_symbols)
+    text_lower = text_one_space.lower()
+    text_list = text_lower.split(" ")
+    helper_list = []
+
+    for word in text_list:
+        if word in helper_list:
```
Comment (chloenott) [inline python/code_challenges/hashtable_repeated_word.py:12]: The time complexity of line 12 is O(n) per [https://wiki.python.org/moin/TimeComplexity](https://wiki.python.org/moin/TimeComplexity). This is inside a for loop (line 11). The combined time complexity is potentially very large!

The workaround to speed things up is to use a set/hashmap/dictionary for `helper_list`; all of these are O(1) lookups when checking if a key exists in helper_list. O(1) inside an O(n) loop is substantially faster!

File: python/docs/hashtable_repeated_word/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+- split string into list using spaces as the separator to get word list
+- create additional empty helper list
+- use for loop to iterate through primary word list
+  - conditional statement each iteration to check if list element is in helper list
+    - if it is, then the element is returned
+    - if it is not, then the element is appended to helper list
+
+### Efficiency
+
+- Time: **O(n*(1/2)(n-1))**, where n is the length of the input in terms of individual words in the string, and (1/2)(n-1) represents the average length of the helper list which increments from initial value of 0 to, at most, value of n-1.
```
Comment (chloenott) [inline python/docs/hashtable_repeated_word/README.md:28]: I like the analysis detail! Makes it clear what you're thinking.

May want to simplify the scaling constants and smaller order terms out. Example:

$n(1/2)(n-1)  \rightarrow   \frac{n(n-1)}{2}  \rightarrow  \frac{n^2}{2}-\frac{n}{2}  \rightarrow O(n^2)$

**Review by chloenott — COMMENTED** [review]
Looks great to me! All tests pass, code looks good. Ideally, a hashmap should be used but I think if you read through the comments below, you'll have a good understanding as to why it's preferred.

File: python/docs/hashtable_repeated_word/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+- Space: **O(2n(n-1))**, input is of n length in terms of individual words in string
```
Comment (chloenott) [inline python/docs/hashtable_repeated_word/README.md:29]: Forgot to comment on the space complexity. $2n(n-1)$ is O(n^2), which is very large. Consider how much space the variables in 5-9 will use, and how that scales with the input. Does the combined size of those variables grow at the same rate as the input, aside from some multiplier?
